server/src/main/java/org/elasticsearch/cluster/features/NodeFeaturesFixupListener.java

```diff
@@ -0,0 +1,217 @@
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+
+@UpdateForV9    // this can be removed in v9
+public class NodeFeaturesFixupListener implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(NodeFeaturesFixupListener.class);
+
+    private static final TimeValue RETRY_TIME = TimeValue.timeValueSeconds(30);
+
+    private final MasterServiceTaskQueue<NodesFeaturesTask> taskQueue;
+    private final ClusterAdminClient client;
+    private final Scheduler scheduler;
+    private final Executor executor;
+    private final Set<String> pendingNodes = Collections.synchronizedSet(new HashSet<>());
+
+    public NodeFeaturesFixupListener(ClusterService service, ClusterAdminClient client, ThreadPool threadPool) {
+        // there tends to be a lot of state operations on an upgrade - this one is not time-critical,
+        // so use LOW priority. It just needs to be run at some point after upgrade.
+        this(
+            service.createTaskQueue("fix-node-features", Priority.LOW, new NodesFeaturesUpdater()),
+            client,
+            threadPool,
+            threadPool.executor(ThreadPool.Names.CLUSTER_COORDINATION)
+        );
+    }
+
+    NodeFeaturesFixupListener(
+        MasterServiceTaskQueue<NodesFeaturesTask> taskQueue,
+        ClusterAdminClient client,
+        Scheduler scheduler,
+        Executor executor
+    ) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+        this.scheduler = scheduler;
+        this.executor = executor;
+    }
+
+    class NodesFeaturesTask implements ClusterStateTaskListener {
+        private final Map<String, Set<String>> results;
+        private final int retryNum;
+
+        NodesFeaturesTask(Map<String, Set<String>> results, int retryNum) {
+            this.results = results;
+            this.retryNum = retryNum;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            logger.error("Could not apply features for nodes {} to cluster state", results.keySet(), e);
+            scheduleRetry(results.keySet(), retryNum);
+        }
+
+        public Map<String, Set<String>> results() {
+            return results;
+        }
+    }
+
+    static class NodesFeaturesUpdater implements ClusterStateTaskExecutor<NodesFeaturesTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodesFeaturesTask> context) {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            var existingFeatures = builder.nodeFeatures();
+
+            boolean modified = false;
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    // double check there are still no features for the node
+                    if (existingFeatures.getOrDefault(e.getKey(), Set.of()).isEmpty()) {
```

Comment:
Why is this needed?
Should not we merge the maps in any case?

Reply:
This is more being extra-safe - in the case the node has left and re-joined in the meantime it will have the features in there already, which may be different to those that were queried before the node was shut down.